dashboard/agent.py

```diff
@@ -92,10 +86,83 @@ async def _check_parent():
                     dashboard_consts.
                     DASHBOARD_AGENT_CHECK_PARENT_INTERVAL_SECONDS)
 
+        check_parent_task = asyncio.create_task(_check_parent())
+
+        # Create an aioredis client for all modules.
+        try:
+            self.aioredis_client = await dashboard_utils.get_aioredis_client(
+                self.redis_address, self.redis_password,
+                dashboard_consts.CONNECT_REDIS_INTERNAL_SECONDS,
+                dashboard_consts.RETRY_REDIS_CONNECTION_TIMES)
+        except (socket.gaierror, ConnectionRefusedError):
+            logger.error(
+                "Dashboard agent suicide, "
+                "Failed to connect to redis at %s", self.redis_address)
+            sys.exit(-1)
+
+        # Create a http session for all modules.
+        self.http_session = aiohttp.ClientSession(
+            loop=asyncio.get_event_loop())
+
+        # Start a grpc asyncio server.
+        await self.server.start()
+
         modules = self._load_modules()
-        await asyncio.gather(_check_parent(),
+
+        # Http server should be initialized after all modules loaded.
+        app = aiohttp.web.Application()
+        app.add_routes(routes=routes.bound_routes())
+
+        # Enable CORS on all routes.
+        cors = aiohttp_cors.setup(
+            app,
+            defaults={
+                "*": aiohttp_cors.ResourceOptions(
+                    allow_credentials=True,
+                    expose_headers="*",
+                    allow_methods="*",
+                    allow_headers=("Content-Type", "X-Header"),
+                )
+            })
+        for route in list(app.router.routes()):
+            cors.add(route)
+
+        runner = aiohttp.web.AppRunner(app)
+        await runner.setup()
+        site = aiohttp.web.TCPSite(runner, self.ip, 0)
+        await site.start()
+        http_host, http_port = site._server.sockets[0].getsockname()
+        logger.info("Dashboard agent http address: %s:%s", http_host,
+                    http_port)
+
+        # Dump registered http routes.
+        dump_routes = [
+            r for r in app.router.routes() if r.method != hdrs.METH_HEAD
+        ]
+        for r in dump_routes:
+            logger.info(r)
+        logger.info("Registered %s routes.", len(dump_routes))
+
+        # Write the dashboard agent port to redis.
+        await self.aioredis_client.set(
```

Comment:
We should remove this from Redis when agents are shutdown. One of the most common workflow of Ray is using autoscaling clusters (and we will invest lots of resources on this), and this means agents will be killed occasionally from the cluster.

Reply:
The agent will be restarted after killed, and the new value will be set to Redis. So, can we leave this from Redis when agents are shutdown?

Reply:
No, If a node is removed from the cluster, the entry for the `agent` will not be cleaned up. `Dashboard` use the `node ip` to access agent entry, so the dead agent entry is not a problem.  If we really need to cleanup the dead agent entry, we can:
- cleanup the agent entry by GCS. (node can't ensure the cleanup)
- use TTL for the agent entry, agent will update the TTL periodically.

Reply:
I think this is not a big deal for now if the only concern is having unused entries to GCS.